File: absactivity/src/main/java/com/sovegetables/BaseFragment.java
package com.sovegetables;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.sovegetables.topnavbar.ITopBarAction;
import com.sovegetables.topnavbar.TopBar;

public abstract class BaseFragment extends Fragment {

    private static IContentViewDelegate sDefaultContentViewDelegate;

    @NonNull
    protected ITopBarAction topBarAction;

    public static void setsDefaultContentViewDelegate(IContentViewDelegate defaultContentViewDelegate) {
        sDefaultContentViewDelegate = defaultContentViewDelegate;
    }

    @Nullable
    @Override
    public final View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View contentView = onBaseCreateView(inflater, container, savedInstanceState);
        sDefaultContentViewDelegate = getContentViewDelegate();
        if(sDefaultContentViewDelegate == null){
            sDefaultContentViewDelegate = new DefaultIContentViewDelegate();
        }
        View realContentView = sDefaultContentViewDelegate.onCreateContentView(contentView);
        if(realContentView == null){
            throw new IllegalArgumentException("IContentViewDelegate's onCreateContentView method can't return null !!!");
        }
        topBarAction = sDefaultContentViewDelegate.onCreateTopBarAction();
        if(topBarAction == null){
            throw new IllegalArgumentException("IContentViewDelegate's onCreateTopBarAction method can't return null !!!");
        }
        topBarAction.setUpTopBar(getTopBar());
        return realContentView;
    }

    protected TopBar getTopBar(){
        return TopBar.NO_ACTION_BAR;
    }

    protected IContentViewDelegate getContentViewDelegate() {
        return null;
    }

    protected View onBaseCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
        return null;
    }
}

File: absactivity/src/main/java/com/sovegetables/BaseFragment.kt
package com.sovegetables

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import com.sovegetables.topnavbar.ITopBarAction
import com.sovegetables.topnavbar.TopBar

abstract class BaseFragment : Fragment(), IEmptyController, ILoadingDialogController, ILoadingController{

    protected var topBarAction: ITopBarAction? = null
    protected var loadingDialogController: ILoadingDialogController? = null
    protected var loadingController: ILoadingController? = null
    protected var emptyController: IEmptyController? = null
    private var contentViewDelegate: IContentView? = null
    private var realContentView: View? = null

    @CallSuper
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?): View? {
        if(realContentView == null){
            val view = onBaseCreateView(inflater, container, savedInstanceState)
            if(view != null){
                contentViewDelegate = getContentViewDelegate()
                if (contentViewDelegate == null) {
                    contentViewDelegate = DefaultIContentView()
                }
                realContentView = contentViewDelegate!!.onCreateContentView(view)
                topBarAction = contentViewDelegate!!.onCreateTopBarAction()
                topBarAction!!.setUpTopBar(getTopBar())
                loadingDialogController = contentViewDelegate!!.getLoadingDialogController()
                emptyController = contentViewDelegate!!.getEmptyController()
                loadingController = contentViewDelegate!!.getLoadingController()
            }
        }
        return realContentView
    }

    open fun getTopBar(): TopBar? {
        return TopBar.NO_ACTION_BAR
    }

    open fun getContentViewDelegate(): IContentView? {
        return null
    }

    open fun addViewBelowTopBar(view: View) {
        if(lifecycle.currentState.isAtLeast(Lifecycle.State.INITIALIZED)){
            contentViewDelegate?.addViewBelowTopBar(view)
        }
    }

    open fun addViewBelowTopBar(@LayoutRes layoutRes: Int) {
        if(lifecycle.currentState.isAtLeast(Lifecycle.State.INITIALIZED)){
            contentViewDelegate?.addViewBelowTopBar(layoutRes)
        }
    }

    open fun onBaseCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(layoutId(), container, false)
    }

    @LayoutRes open fun layoutId(): Int{
        return 0
    }

    override fun showEmpty(msg: String?, icon: Int, model: IEmptyController.Model?) {
        emptyController?.showEmpty(msg, icon, model)
    }

    override fun hideEmpty() {
        emptyController?.hideEmpty()
    }

    override fun showLoadingDialog(
        msg: String?,
        icon: Int,
        canceled: Boolean,
        model: ILoadingDialogController.Model?
    ) {
        loadingDialogController?.showLoadingDialog(msg, icon, canceled, model)
    }

    override fun hideLoadingDialog() {
        loadingDialogController?.hideLoadingDialog()
    }

    override fun isLoadingDialogShow(): Boolean {
        return loadingDialogController?.isLoadingDialogShow()?:false
    }

    override fun showLoading() {
        loadingController?.showLoading()
    }

    override fun hideLoading() {
        loadingController?.hideLoading()
    }

    override fun isLoadingShow(): Boolean {
        return loadingController?.isLoadingShow()?:false
    }
}